source/interact-data/transaction.txt

```diff
@@ -0,0 +1,324 @@
+   :start-after: start-commit-abort
+   :end-before: end-commit-abort
+   :language: ruby
+   :dedent:
+
+.. note::
+
+   If a session ends and includes an open transaction, the transaction is
+   automatically ended.
+
+You can retry the transaction commit if it fails initially. The
+following example demonstrates how to retry the transaction when {+odm+}
+raises the ``UNKNOWN_TRANSACTION_COMMIT_RESULT_LABEL`` exception:
+
+.. literalinclude:: /includes/interact-data/transaction.rb
+   :start-after: start-commit-retry
+   :end-before: end-commit-retry
+   :language: ruby
+   :dedent:
+
+Client Behavior
+~~~~~~~~~~~~~~~
+
+To perform operations within a transaction, operations must use the same
+client that the session was initiated on. By default, all operations
+are performed by using the default client.
+
+To explicitly use a different client, use the ``with()`` method: 
+
+.. literalinclude:: /includes/interact-data/transaction.rb
+   :start-after: start-other-client
+   :end-before: end-other-client
+   :language: ruby
+   :dedent:
+
+.. _mongoid-txn-session:
+
+Session API
+-----------
+
+You can use sessions in {+odm+} in a similar way that you can
+perform a transaction. You can call the ``with_session()`` method on a
+model class or on an instance of a model and perform some operations in
+a block. All operations in the block will be performed in the context of
+single session.
+
+The following limitations apply when using sessions:
+
+- You cannot share a session across threads. Sessions are not thread-safe.
+
+- You cannot nest sessions. For example, you cannot call the ``with_session()``
+  method on a model class within the block passed to
+  the ``with_session()`` method on another model class. The following
+  code demonstrates a nested session that results in an error:
+
+  .. code-block:: ruby
+
+     Book.with_session(causal_consistency: true) do
+       # Nesting sessions results in errors
+       Film.with_session(causal_consistency: true) do
+         ...
+       end
+     end
+
+- All model classes and instances used within the session block must use
+  the same driver client. For example, if you specify different
+  a different client for a model used in the block than those of the
+  model or instance that you called ``with_session()`` on, {+odm+} returns
+  an error.
+
+Examples
+~~~~~~~~
+
+You can use the ``with_session()`` method on a model class and pass it session
+options to perform a block of operations in the context of a session.
+
+The following code enables the ``causal_consistency`` option to
+guarantee the order of operations when creating a session on the
+``Book`` model, then performs data operations:
+
+.. literalinclude:: /includes/interact-data/transaction.rb
+   :start-after: start-model-session
+   :end-before: end-model-session
+   :language: ruby
+   :dedent:
+
+To learn more about the available session options, see the
+:ruby-api:`Session class constructor details
```

Comment:
Note: API links are broken until the next snooty parser release see https://github.com/mongodb/snooty-parser/pull/633